packages/webapp/src/_app/eos/secret-publisher.ts

```diff
@@ -0,0 +1,217 @@
+    console.info(publisherKey, recipientKeys);
+
+    // TODO: check publisherkey matches the current one in browser
+
+    const transientKeyPair = await generateRandomTransientKey();
+    console.info(transientKeyPair);
+
+    const keks = await ecdhRecipientsKeyEncriptionKeys(
+        [publisherKey, ...recipientKeys],
+        transientKeyPair.privateKey
+    );
+    console.info(keks);
+
+    const sessionKey = await generateRandomSessionKey();
+    console.info(sessionKey);
+
+    const encryptedSessionKeys = await encryptSessionKeys(sessionKey, keks);
+    console.info(encryptedSessionKeys);
+    const encryptedMessage = await encryptMessage(sessionKey, message);
+    console.info(encryptedMessage);
+
+    const obj = {
+        encryptedSessionKeys,
+        recipientKeys,
+        transientKeyPair,
+        encryptedMessage,
+    };
+    console.info(obj);
+    return obj;
+};
+
+const fetchRecipientKeys = async (accounts: string[]) => {
+    // TODO: real implementation
+    // fetch eden.keys_table( where account = each account item )
+
+    const mockedKeysLen = accounts.length;
+    const mockedKeys = [];
+    for (let i = 0; i < mockedKeysLen; i++) {
+        mockedKeys.push(
+            "EOS5wGY8RcG8PuGYTPU9QwmiEm9XReBkK4SGtSuS4f1oeX4TNGaHT"
+        );
+        // temp PK: 5KcMypBPGByXbeBphsVzyyzUvg31tWNKsgrfXs2MmanYgYf4gao
+    }
+    return mockedKeys;
+};
+
+const generateRandomTransientKey = async () => {
+    const privateKey = await ecc.randomKey();
+    const publicKey = ecc.privateToPublic(privateKey);
+    return { privateKey, publicKey };
+};
+
+const ecdhRecipientsKeyEncriptionKeys = async (
+    recipientPublicKeys: string[],
+    transientPrivateKeyString: string
+) => {
+    const transientPrivateKey = PrivateKey.fromString(
+        transientPrivateKeyString
+    );
+    const transientEc = ec.keyPair({
+        priv: new bn(transientPrivateKey.d.toHex(), 16),
+    });
+    return await Promise.all(
+        recipientPublicKeys.map(async (publicKey: string) => {
+            const recipientEcPublicKey = ec.keyFromPublic(
+                PublicKey.fromString(publicKey).toHex(),
+                "hex"
+            );
+            const ecdhSecret = transientEc.derive(
+                recipientEcPublicKey.getPublic()
+            );
+            return await hkdfSha256FromEcdh(
+                ecdhSecret.toArrayLike(ArrayBuffer) // TODO: review endianness, empty = Big Endian
+            );
+        })
+    );
+};
+
+const hkdfSha256FromEcdh = async (ecdhSecret: ArrayBuffer) => {
+    const sharedSecretKey = await crypto.subtle.importKey(
+        "raw",
+        ecdhSecret,
+        { name: "HKDF" },
+        false,
+        ["deriveKey", "deriveBits"]
+    );
+    const derivedKey = await crypto.subtle.deriveBits(
+        {
+            name: "HKDF",
+            hash: "SHA-256",
+            salt: new Uint8Array([]),
+            info: new Uint8Array([]),
+        },
+        sharedSecretKey,
+        256
```

Comment:
This will end up generating an AES-256 key.  It's not a security problem, but it is a bit wasteful as we only have 128-bit security through the whole system.  Also, I think it would be simpler to use deriveKey and avoid having to importKey.

Reply:
Yeah, I think in general I can just stop exporting and re-importing keys, I was going to do that in my final refactor! thanks for the 128-bit heads up. Will change it!